### ISSUE TITLE
Is there the ability to access previous version info in replacement template?

### ISSUE THREAD

**Issue by hartzell** [lgiordani/punch#43]
TL;DR -- I'd like to be able to add an "Unreleased" section at the top of a CHANGELOG.md file when I increment a "-dev" part. 
 I think that I have a small change that should do it, but Jinja complains about it.  Perhaps there's a way to make Jinja happy or perhaps there's a better solution to my problem?

---

I'm using the technique described in the [Complex Serializers section of the Advanced page](https://punch.readthedocs.io/en/latest/advanced/) to replace an "Unreleased" string with a version number in my `CHANGELOG.md`.

I'm also using a variant of the `""`/`alpha`/`beta` technique described in the [SemVer with prerelease metadata section of the examples page](https://punch.readthedocs.io/en/latest/examples/) to manage a `-dev` suffix for the version info.

If I were able to access the previous version information in the replacement template, then I could insert a new blank section at the top of the `CHANGELOG.md` when I bump the dev part from `""` to `"dev"`.

The flow would go something like:

- current contents of the CHANGELOG.md

  ```
  # Unreleased

  - new feature
  - another new feature
  ```

- run `punch --part minor`, changes CHANGELOG.md to

  ```
  # 1.3.0

  - new feature
  - another new feature
  ```

- run `punch --part dev` (which adds `-dev` to the version info in other files), which updates the CHANGELOG.md to:

  ```
  # Unreleased

  # 1.3.0

  - new feature
  - another new feature
  ```

I think it would work with a search rule that just searches the current version (including the `# `) and a replacement template like:

```
'''
# Unreleased

# {{prev.major}}.{{prev.minor}}.{{prev.patch}} -- {{prev.year}}-{{prev.month}}-{{prev.day}}
'''
```

I've played around with a diff like so:

```diff
diff --git a/punch/cli.py b/punch/cli.py
index 84badfd..e5c20f7 100644
--- a/punch/cli.py
+++ b/punch/cli.py
@@ -310,9 +310,11 @@ def main(original_args=None):
             exc
         )

+    new_values=new_version.as_dict()
+    new_values.update({"prev": current_version.as_dict()})
     changes = global_replacer.run_all_serializers(
         current_version.as_dict(),
-        new_version.as_dict()
+        new_values
     )

     if args.verbose:
```

but I'm ending up with:

```
Traceback (most recent call last):
  File "/local_scratch/george.hartzell/tmp/ngs-ingress/.venv/bin/punch", line 8, in <module>
    sys.exit(main())
  File "/local_scratch/george.hartzell/tmp/ngs-ingress/.venv/lib/python3.7/site-packages/punch/cli.py", line 316, in main
    changes = global_replacer.run_all_serializers(
  File "/local_scratch/george.hartzell/tmp/ngs-ingress/.venv/lib/python3.7/site-packages/punch/file_updater.py", line 37, in update
    new_version
  File "/local_scratch/george.hartzell/tmp/ngs-ingress/.venv/lib/python3.7/site-packages/punch/replacer.py", line 88, in replace
    new_version_dict
  File "/local_scratch/george.hartzell/tmp/ngs-ingress/.venv/lib/python3.7/site-packages/punch/replacer.py", line 78, in run_all_serializers
    name, current_version_dict, new_version_dict)
  File "/local_scratch/george.hartzell/tmp/ngs-ingress/.venv/lib/python3.7/site-packages/punch/replacer.py", line 63, in run_serializer
    replace_template.render(**new_version_dict)
  File "/local_scratch/george.hartzell/tmp/ngs-ingress/.venv/lib/python3.7/site-packages/jinja2/environment.py", line 1090, in render
    self.environment.handle_exception()
  File "/local_scratch/george.hartzell/tmp/ngs-ingress/.venv/lib/python3.7/site-packages/jinja2/environment.py", line 832, in handle_exception
    reraise(*rewrite_traceback_stack(source=source))
  File "/local_scratch/george.hartzell/tmp/ngs-ingress/.venv/lib/python3.7/site-packages/jinja2/_compat.py", line 28, in reraise
    raise value.with_traceback(tb)
  File "<template>", line 3, in top-level template code
  File "/local_scratch/george.hartzell/tmp/ngs-ingress/.venv/lib/python3.7/site-packages/jinja2/environment.py", line 471, in getattr
    return getattr(obj, attribute)
jinja2.exceptions.UndefinedError: 'prev' is undefined
```

Any suggestions to get me unstuck, or alternate ways of reaching my goal?

**Comment by lgiordani** [comment]
Hey @hartzell thanks for the issue, I will have a look asap

**Comment by hartzell** [comment]
I've figured out how to do this with no modifications to punch.  I'm using git-flow-esque workflow and a semantic-ish versioning scheme with the addition that the "version" of the unreleased code on my `develop` branch has a `-dev` suffix.  Making a release involves:

1. Making a "release" branch from the `develop` branch.
1. Running `punch --part minor` (or `patch`, or `major`, as appropriate) which updates:
   - the various files that contain the version number (e.g. `0.1.0-dev`), bumping the version info 
     and removing the `-dev`; and
   - the changelog, replacing `## Unreleased` with the version info.
2. Touching up the changelog, doing whatever, committing whatever, and merging the release branch to `main`.
3. Back on the "release" branch, running `punch --part dev`, which updates:
   - the various files that contain the version number, appending the `-dev` suffix; and
   - the changelog, adding a new `## Unreleased` section above the last versioned entry.
4. Merging the "release" branch into `develop`.

The `punch_config.py` looks something like this:

```python
__config_version__ = 1

GLOBALS = {
     "serializer": "{{major}}.{{minor}}.{{patch}}{{ '-{}'.format(dev) if dev }}",
}

FILES = [
    "VERSION",
    {
        "path": "Changelog.md",
        "serializer": {
            "semver": {
                "search": "## Unreleased",
                "replace": "## [{{major}}.{{minor}}.{{patch}}] -- {{year}}-{{month}}-{{day}}",
            },
            # By the time this rule runs when bumping the 'dev'
            # field, the GLOBAL serializer already slapped -dev onto
            # the line in the CHANGELOG, so take it into account
            "unreleased": {
               "search": "## [{{major}}.{{minor}}.{{patch}}-dev] -- {{year}}-{{month}}-{{day}}",
               "replace":"""## Unreleased

## [{{major}}.{{minor}}.{{patch}}] -- {{year}}-{{month}}-{{day}}
"""
,
           },
        },
    },
]

VERSION = [
    "major",
    "minor",
    "patch",
    {"name": "dev", "type": "value_list", "allowed_values": ["", "dev"],},
    {"name": "year", "type": "date", "fmt": "YYYY"},
    {"name": "month", "type": "date", "fmt": "MM"},
    {"name": "day", "type": "date", "fmt": "DD"},
]
```

`punch_version.py` looks something like this:

```python
major = 1
minor = 0
patch = 0
dev = 'dev'
year = '1970'
month = '1'
day = '1'
```

a `VERSION` file might look like so:

```
1.0.0-dev
```

and finally a `Changelog.md` might look like so:

```markdown
# Changelog

## Unreleased

## [1.0.0] -- 1970-1-1

Changes go here
```

**Comment by lgiordani** [comment]
George, many thanks and sorry for not coming back to you earlier in this busy time. I'm glad you sorted it out, but I still think your use case is interesting, and I will keep this in mind for future changes. In the meantime, I wish you a happy Christmas!
